src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -440,7 +481,11 @@ public void DoWriteIfNeeded()
                 }, this);
 
                 Self._head = _lockedEnd.Block;
-                Self._head.Start = _lockedEnd.Index;
```

Comment:
@benaadams You don't want to move the set above the write in case write throws. If that's the case, we still want to ensure that the blocks get returned to the pool eventually.

Even if the write callback gets executed inline `Self._head` should never be null after `Self._head = _lockedEnd.Block`. We already know `_lockedEnd.Block` can't be null because `LockWrite` would exit early leaving `ByteCount` at zero.